File: sdk/typescript/src/imessage/resources/imessages.ts
/**
 * inkbox-imessage/resources/imessages.ts
 *
 * iMessage operations: dedicated numbers, send, list, conversations,
 * reactions, read receipts, typing indicators, media upload.
 *
 * Conversation operations key off `conversationId` / `agentIdentityId`.
 * Dedicated number inventory is managed through `listNumbers` and
 * `claimNumber`.
 */

import { HttpTransport, validateIdempotencyKey } from "../../_http.js";
import {
  IMessage,
  IMessageAssignment,
  IMessageConversation,
  IMessageConversationSummary,
  IMessageDedicatedNumberType,
  IMessageNumber,
  IMessageMarkReadResult,
  IMessageMediaUpload,
  IMessageReaction,
  IMessageReactionType,
  IMessageSendStyle,
  IMessageTriageNumber,
  RawIMessage,
  RawIMessageAssignment,
  RawIMessageConversation,
  RawIMessageConversationSummary,
  RawIMessageNumber,
  RawIMessageReaction,
  RawIMessageTriageNumber,
  parseIMessage,
  parseIMessageAssignment,
  parseIMessageConversation,
  parseIMessageConversationSummary,
  parseIMessageNumber,
  parseIMessageReaction,
  parseIMessageTriageNumber,
} from "../types.js";

const SENDABLE_REACTIONS = new Set<string>([
  "love",
  "like",
  "dislike",
  "laugh",
  "emphasize",
  "question",
  "eyes",
]);

function validateSendableReaction(reaction: IMessageReactionType | string): string {
  if (!SENDABLE_REACTIONS.has(reaction)) {
    throw new Error(
      `reaction must be one of: ${[...SENDABLE_REACTIONS].join(", ")}`,
    );
  }
  return reaction;
}

export class IMessagesResource {
  constructor(private readonly http: HttpTransport) {}

  /**
   * Return the active triage line and the connect command.
   *
   * Recipients text the returned `connectCommand` (e.g.
   * `connect @your-handle`) to the triage `number` to get connected to
   * an agent identity. Resolve this at runtime instead of hardcoding
   * the number — the line can change.
   *
   * @throws {InkboxAPIError} 404 when no triage line is active.
   */
  async getTriageNumber(): Promise<IMessageTriageNumber> {
    const data = await this.http.get<RawIMessageTriageNumber>("/triage-number");
    return parseIMessageTriageNumber(data);
  }

  /**
   * List every non-released dedicated iMessage number owned by the
   * organization, including unattached numbers.
   */
  async listNumbers(): Promise<IMessageNumber[]> {
    const data = await this.http.get<RawIMessageNumber[]>("/numbers");
    return data.map(parseIMessageNumber);
  }

  /**
   * Claim one dedicated iMessage number for the organization.
   *
   * Claiming does not attach the number to an identity. Pass
   * `imessageNumberType` during identity creation or update to claim and
   * attach atomically, or pass an owned number's id as `imessageNumberId`
   * during identity update.
   *
   * Reuse the same caller-generated `idempotencyKey` when retrying an
   * ambiguous request. A new key can claim another number.
   *
   * @throws {DedicatedIMessageNumberQuotaExceededError} 402 when the
   *   organization has reached its quota for the requested number type.
   * @throws {DedicatedIMessageNumberInventoryPendingError} 503 when number
   *   inventory is pending; inspect `retryAfterSeconds` before retrying.
   * @throws {IdempotencyKeyReusedError} 409 when the key was already used
   *   for a different request.
   */
  async claimNumber(options: {
    type: IMessageDedicatedNumberType;
    idempotencyKey: string;
  }): Promise<IMessageNumber> {
    validateIdempotencyKey(options.idempotencyKey);
    const data = await this.http.post<RawIMessageNumber>("/numbers", {
      type: options.type,
    }, {
      headers: { "Idempotency-Key": options.idempotencyKey },
    });
    return parseIMessageNumber(data);
  }

  /**
   * Send an outbound iMessage through an existing assignment.
   *
   * Shared and dedicated-inbound numbers require the recipient to connect
   * first. An identity attached to a dedicated-outbound number may initiate
   * a conversation, subject to server-side consent and rate limits.
   * Inbound replies and reactions arrive via identity-owned webhook
   * subscriptions (`inkbox.webhooks.subscriptions.create({
   * agentIdentityId, url, eventTypes: ["imessage.received", ...] })`).
   *
   * @param options.to - One E.164 recipient or 1–8 distinct recipients. Two
   *   or more recipients select or create a dedicated-outbound group.
   *   Mutually exclusive with `conversationId`.
   * @param options.conversationId - Existing conversation UUID to reply into.
   * @param options.text - Message body.
   * @param options.mediaUrls - Media URLs (at most one). Use
   *   {@link uploadMedia} to turn raw bytes into a sendable URL first.
   * @param options.sendStyle - Optional expressive send style. The same
   *   `IMessageSendStyle` values work for one-to-one sends, new groups, and
   *   replies by group `conversationId`, including sends with one media URL.
   * @param options.agentIdentityId - Identity to send as. Required for
   *   org-wide API keys when sending by `to`; ignored for
   *   identity-scoped keys (the key's identity wins).
   *
   * @throws {InkboxAPIError} 400 when the identity is not
   *   iMessage-enabled; 403 when the recipient is blocked by a contact
   *   rule.
   */
  async send(options: {
    to?: string | string[] | null;
    conversationId?: string | null;
    text?: string | null;
    mediaUrls?: string[] | null;
    sendStyle?: IMessageSendStyle | string | null;
    agentIdentityId?: string | null;
  }): Promise<IMessage> {
    const body: {
      to?: string | string[];
      conversation_id?: string;
      text?: string;
      media_urls?: string[];
      send_style?: string;
    } = {};
    if (options.to != null) {
      body.to = options.to;
    }
    if (options.conversationId != null) {
      body.conversation_id = options.conversationId;
    }
    if (options.text != null) {
      body.text = options.text;
    }
    if (options.mediaUrls != null) {
      body.media_urls = options.mediaUrls;
    }
    if (options.sendStyle != null) {
      body.send_style = options.sendStyle;
    }
    const params: Record<string, string> = {};
    if (options.agentIdentityId != null) {
      params["agent_identity_id"] = options.agentIdentityId;
    }
    const data = await this.http.post<{ message: RawIMessage }>(
      "/messages",
      body,
      { params },
    );
    return parseIMessage(data.message);
  }

  /**
   * List iMessages visible to the caller, newest first.
   *
   * Identity-scoped API keys never see contact-rule-blocked rows
   * regardless of `isBlocked` (filtered server-side). Admin/JWT callers
   * see everything by default.
   *
   * @param options.agentIdentityId - Narrow to one agent identity.
   *   Ignored for identity-scoped keys (always their own identity).
   * @param options.conversationId - Narrow to one conversation.
   * @param options.limit - Max results (1–200). Defaults to 50.
   * @param options.offset - Pagination offset. Defaults to 0.
   * @param options.isRead - Filter by read state.
   * @param options.isBlocked - Tri-state filter. `true` for only blocked,
   *   `false` for only non-blocked, omit for all.
   */
  async list(
    options?: {
      agentIdentityId?: string;
      conversationId?: string;
      limit?: number;
      offset?: number;
      isRead?: boolean;
      isBlocked?: boolean;
      includeGroups?: boolean;
      startDatetime?: string;
      endDatetime?: string;
      tz?: string;
    },
  ): Promise<IMessage[]> {
    const params: Record<string, string | number | boolean> = {
      limit: options?.limit ?? 50,
      offset: options?.offset ?? 0,
    };
    if (options?.agentIdentityId !== undefined) {
      params["agent_identity_id"] = options.agentIdentityId;
    }
    if (options?.conversationId !== undefined) {
      params["conversation_id"] = options.conversationId;
    }
    if (options?.isRead !== undefined) {
      params["is_read"] = options.isRead;
    }
    if (options?.isBlocked !== undefined) {
      params["is_blocked"] = options.isBlocked;
    }
    if (options?.includeGroups === true) {
      params["include_groups"] = true;
    }
    if (options?.startDatetime !== undefined) params["start_datetime"] = options.startDatetime;
    if (options?.endDatetime !== undefined) params["end_datetime"] = options.endDatetime;
    if (options?.tz !== undefined) params["tz"] = options.tz;
    const data = await this.http.get<RawIMessage[]>("/messages", params);
    return data.map(parseIMessage);
  }

  /**
   * List active iMessage connections, newest first.
   *
   * One row per recipient currently connected to an agent identity
   * through triage. Released connections are not returned.
   *
   * @param options.agentIdentityId - Narrow to one agent identity.
   *   Ignored for identity-scoped keys (always their own identity).
   * @param options.limit - Max results (1–200). Defaults to 50.
   * @param options.offset - Pagination offset. Defaults to 0.
   */
  async listAssignments(
    options?: {
      agentIdentityId?: string;
      limit?: number;
      offset?: number;
    },
  ): Promise<IMessageAssignment[]> {
    const params: Record<string, string | number> = {
      limit: options?.limit ?? 50,
      offset: options?.offset ?? 0,
    };
    if (options?.agentIdentityId !== undefined) {
      params["agent_identity_id"] = options.agentIdentityId;
    }
    const data = await this.http.get<RawIMessageAssignment[]>("/assignments", params);
    return data.map(parseIMessageAssignment);
  }

  /**
   * List iMessage conversations with latest-message preview.
   *
   * @param options.agentIdentityId - Narrow to one agent identity.
   *   Ignored for identity-scoped keys (always their own identity).
   * @param options.limit - Max results (1–200). Defaults to 50.
   * @param options.offset - Pagination offset. Defaults to 0.
   * @param options.isBlocked - Tri-state filter applied to the
   *   underlying messages. `true` for only blocked, `false` for only
   *   non-blocked, omit for all.
   */
  async listConversations(
    options?: {
      agentIdentityId?: string;
      limit?: number;
      offset?: number;
      isBlocked?: boolean;
      includeGroups?: boolean;
      startDatetime?: string;
      endDatetime?: string;
      tz?: string;
    },
  ): Promise<IMessageConversationSummary[]> {
    const params: Record<string, string | number | boolean> = {
      limit: options?.limit ?? 50,
      offset: options?.offset ?? 0,
    };
    if (options?.agentIdentityId !== undefined) {
      params["agent_identity_id"] = options.agentIdentityId;
    }
    if (options?.isBlocked !== undefined) {
      params["is_blocked"] = options.isBlocked;
    }
    if (options?.includeGroups === true) {
      params["include_groups"] = true;
    }
    if (options?.startDatetime !== undefined) params["start_datetime"] = options.startDatetime;
    if (options?.endDatetime !== undefined) params["end_datetime"] = options.endDatetime;
    if (options?.tz !== undefined) params["tz"] = options.tz;
    const data = await this.http.get<RawIMessageConversationSummary[]>(
      "/conversations",
      params,
    );
    return data.map(parseIMessageConversationSummary);
  }

  /**
   * Get one iMessage conversation by ID.
   *
   * @param conversationId - UUID of the conversation.
   * @param options.agentIdentityId - Optional identity assertion; 404s
   *   when the conversation belongs to a different identity.
   */
  async getConversation(
    conversationId: string,
    options?: { agentIdentityId?: string },
  ): Promise<IMessageConversation> {
    const params: Record<string, string> = {};
    if (options?.agentIdentityId !== undefined) {
      params["agent_identity_id"] = options.agentIdentityId;
    }
    const data = await this.http.get<RawIMessageConversation>(
      `/conversations/${conversationId}`,
      params,
    );
    return parseIMessageConversation(data);
  }

  /**
   * Send a tapback reaction to an inbound one-to-one or group message.
   *
   * @param options.messageId - UUID of the message being reacted to.
   * @param options.reaction - Tapback kind. Sends accept `love`, `like`,
   *   `dislike`, `laugh`, `emphasize`, `question`, and `eyes`; `custom` is
   *   inbound-only and rejected locally.
   * @param options.partIndex - Part of a multi-part message to react to.
   *   Defaults to 0.
   * @throws {Error} when `reaction` is custom or not one of the seven
   *   provider-supported named tapbacks.
   */
  async sendReaction(options: {
    messageId: string;
    reaction: IMessageReactionType | string;
    partIndex?: number;
  }): Promise<IMessageReaction> {
    const reaction = validateSendableReaction(options.reaction);
    const data = await this.http.post<RawIMessageReaction>("/reactions", {
      message_id: options.messageId,
      reaction,
      part_index: options.partIndex ?? 0,
    });
    return parseIMessageReaction(data);
  }

  /**
   * Send a one-to-one read receipt and mark inbound messages read locally.
   * Group conversations return 409.
   *
   * @param conversationId - UUID of the conversation.
   * @returns Object with `conversationId` and `updatedCount`.
   */
  async markConversationRead(
    conversationId: string,
  ): Promise<IMessageMarkReadResult> {
    const data = await this.http.post<{
      conversation_id: string;
      updated_count: number;
    }>("/mark-read", { conversation_id: conversationId });
    return {
      conversationId: data.conversation_id,
      updatedCount: data.updated_count,
    };
  }

  /**
   * Show a typing indicator to a one-to-one recipient.
   * Group conversations return 409.
   *
   * @param conversationId - UUID of the conversation.
   */
  async sendTyping(conversationId: string): Promise<void> {
    await this.http.post("/typing", { conversation_id: conversationId });
  }

  /**
   * Upload media and get back a URL usable in `mediaUrls`.
   *
   * @param options.content - Raw file bytes (max 10 MiB).
   * @param options.filename - Original filename, used for type inference.
   * @param options.contentType - Optional MIME type; defaults
   *   server-side to `application/octet-stream`.
   */
  async uploadMedia(options: {
    content: Uint8Array | Blob;
    filename: string;
    contentType?: string;
  }): Promise<IMessageMediaUpload> {
    const data = await this.http.postMultipart<{
      media_url: string;
      content_type?: string | null;
      size?: number | null;
    }>("/media", {
      fieldName: "file",
      filename: options.filename,
      content: options.content,
      contentType: options.contentType ?? "application/octet-stream",
    });
    return {
      mediaUrl: data.media_url,
      contentType: data.content_type ?? null,
      size: data.size ?? null,
    };
  }
}

File: sdk/python/inkbox/imessage/resources/imessages.py
"""
inkbox/imessage/resources/imessages.py

iMessage operations: send, list, conversations, reactions, read
receipts, typing indicators, media upload.

iMessage messaging operations are identity-scoped, so they key off
``conversation_id`` / ``agent_identity_id`` rather than a local number ID.
One-to-one conversations may carry assignment state; groups require a dedicated
outbound number. This resource also lists and claims dedicated numbers.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any
from uuid import UUID

from inkbox.imessage.types import (
    IMessage,
    IMessageAssignment,
    IMessageConversation,
    IMessageConversationSummary,
    IMessageNumber,
    IMessageNumberType,
    IMessageMarkReadResult,
    IMessageMediaUpload,
    IMessageReaction,
    IMessageReactionType,
    IMessageSendStyle,
    IMessageTriageNumber,
    _dedicated_number_type,
    _validate_idempotency_key,
)

if TYPE_CHECKING:
    from inkbox._http import HttpTransport


class IMessagesResource:

    def __init__(self, http: HttpTransport) -> None:
        self._http = http

    def get_triage_number(self) -> IMessageTriageNumber:
        """Return the active triage number and the connect command.

        Recipients text the returned ``connect_command`` (e.g.
        ``connect @your-handle``) to the triage ``number`` to get
        connected to an agent identity. Resolve this at runtime instead
        of hardcoding the number — it can change.

        Raises:
            InkboxAPIError: 404 when no triage number is active.
        """
        data = self._http.get("/triage-number")
        return IMessageTriageNumber._from_dict(data)

    def list_numbers(self) -> list[IMessageNumber]:
        """List the organization's dedicated iMessage numbers.

        Attached and unattached numbers are both returned. Released numbers are
        excluded by the server.
        """
        data = self._http.get("/numbers")
        return [IMessageNumber._from_dict(number) for number in data]

    def claim_number(
        self,
        *,
        type: IMessageNumberType | str,
        idempotency_key: str,
    ) -> IMessageNumber:
        """Claim a new dedicated iMessage number for the organization.

        ``idempotency_key`` must be stable across retries of the same logical
        claim. Dedicated inbound numbers require the recipient to message first.
        Dedicated outbound numbers may start new conversations, subject to
        server-side consent, contact-rule, and rate-limit checks.

        Raises:
            DedicatedIMessageNumberQuotaExceededError: The organization has
                reached its plan limit for the requested number type.
            DedicatedIMessageNumberInventoryPendingError: No matching number is
                currently available; retry after the reported interval.
            IdempotencyKeyReusedError: The key was previously used for
                a different claim.
        """
        number_type = _dedicated_number_type(type).value
        key = _validate_idempotency_key(idempotency_key)
        data = self._http.post(
            "/numbers",
            json={"type": number_type},
            headers={"Idempotency-Key": key},
        )
        return IMessageNumber._from_dict(data)

    def send(
        self,
        *,
        to: str | list[str] | None = None,
        conversation_id: UUID | str | None = None,
        text: str | None = None,
        media_urls: list[str] | None = None,
        send_style: IMessageSendStyle | str | None = None,
        agent_identity_id: UUID | str | None = None,
    ) -> IMessage:
        """Send an outbound iMessage.

        Shared and dedicated inbound service require an existing assignment.
        An identity attached to a dedicated outbound number may start a new
        conversation, subject to server-side policy checks.

        Args:
            to: One E.164 recipient or 1–8 distinct recipients. Two or more
                recipients select or create a dedicated-outbound group.
                Mutually exclusive with ``conversation_id``.
            conversation_id: Existing conversation UUID to reply into.
            text: Message body.
            media_urls: Media URLs (at most one). Pass with ``text`` or
                by themselves. Use :meth:`upload_media` to turn raw
                bytes into a sendable URL first.
            send_style: Optional expressive send style
                (see ``IMessageSendStyle``).
            agent_identity_id: Identity to send as. Required for
                org-wide API keys when sending by ``to``; ignored for
                identity-scoped keys (the key's identity wins).

        Returns:
            The queued ``IMessage`` row. Inbound replies and reactions
            arrive via identity-owned webhook subscriptions
            (``inkbox.webhooks.subscriptions.create(agent_identity_id=...,
            url=..., event_types=["imessage.received", ...])``).

        Raises:
            InkboxAPIError: 400 when the identity is not iMessage-enabled;
                404 when no assignment exists (shared service includes the
                connect command and router number, while dedicated inbound
                directs the recipient to the attached number); 409 when an
                existing conversation is inactive or a recipient-first line
                has not received a message yet, with setup-appropriate next
                steps in the detail; 429 when the identity's rolling 24-hour
                send cap is reached.
            RecipientBlockedError: 403 when the recipient is blocked by a
                contact rule.
        """
        body: dict[str, Any] = {}
        if to is not None:
            body["to"] = to
        if conversation_id is not None:
            body["conversation_id"] = str(conversation_id)
        if text is not None:
            body["text"] = text
        if media_urls is not None:
            body["media_urls"] = media_urls
        if send_style is not None:
            body["send_style"] = (
                send_style.value if isinstance(send_style, IMessageSendStyle) else send_style
            )

        params: dict[str, Any] | None = None
        if agent_identity_id is not None:
            params = {"agent_identity_id": str(agent_identity_id)}
        data = self._http.post("/messages", json=body, params=params)
        return IMessage._from_dict(data["message"])

    def list(
        self,
        *,
        agent_identity_id: UUID | str | None = None,
        conversation_id: UUID | str | None = None,
        limit: int = 50,
        offset: int = 0,
        is_read: bool | None = None,
        is_blocked: bool | None = None,
        include_groups: bool = False,
        start_datetime: str | None = None,
        end_datetime: str | None = None,
        tz: str | None = None,
    ) -> list[IMessage]:
        """List iMessages visible to the caller, newest first.

        Identity-scoped API keys never see contact-rule-blocked rows
        regardless of ``is_blocked`` — the server filters them at the
        access-policy layer. Admin-scoped keys and JWT humans see
        everything by default.

        Args:
            agent_identity_id: Narrow to one agent identity. Ignored
                for identity-scoped keys (always their own identity).
            conversation_id: Narrow to one conversation.
            limit: Max results to return (1–200).
            offset: Pagination offset.
            is_read: Filter by read state (``True``, ``False``, or ``None`` for all).
            is_blocked: Tri-state filter — ``True`` for only blocked,
                ``False`` for only non-blocked, ``None`` for all.
            include_groups: Include group messages. Defaults to ``False`` for
                backwards-compatible one-to-one listings. A specific
                ``conversation_id`` is returned even when this is ``False``.
            start_datetime: Inclusive lower bound on ``created_at`` (str). Bare
                dates resolve to the start of that day; naive datetimes are
                interpreted in ``tz``; zoned datetimes are exact instants.
                ``None`` leaves the range open on this side.
            end_datetime: Upper bound on ``created_at`` (str). A bare date is
                whole-day inclusive. ``None`` leaves the range open.
            tz: IANA timezone name (str) governing zone-less values;
                ``None`` means UTC.
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if agent_identity_id is not None:
            params["agent_identity_id"] = str(agent_identity_id)
        if conversation_id is not None:
            params["conversation_id"] = str(conversation_id)
        if is_read is not None:
            params["is_read"] = is_read
        if is_blocked is not None:
            params["is_blocked"] = is_blocked
        if include_groups:
            params["include_groups"] = True
        if start_datetime is not None:
            params["start_datetime"] = start_datetime
        if end_datetime is not None:
            params["end_datetime"] = end_datetime
        if tz is not None:
            params["tz"] = tz
        data = self._http.get("/messages", params=params)
        return [IMessage._from_dict(m) for m in data]

    def list_assignments(
        self,
        *,
        agent_identity_id: UUID | str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> list[IMessageAssignment]:
        """List active iMessage connections, newest first.

        One row per recipient currently connected to an agent identity
        through triage. Released connections are not returned.

        Args:
            agent_identity_id: Narrow to one agent identity. Ignored
                for identity-scoped keys (always their own identity).
            limit: Max results to return (1–200).
            offset: Pagination offset.
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if agent_identity_id is not None:
            params["agent_identity_id"] = str(agent_identity_id)
        data = self._http.get("/assignments", params=params)
        return [IMessageAssignment._from_dict(a) for a in data]

    def list_conversations(
        self,
        *,
        agent_identity_id: UUID | str | None = None,
        limit: int = 50,
        offset: int = 0,
        is_blocked: bool | None = None,
        include_groups: bool = False,
        start_datetime: str | None = None,
        end_datetime: str | None = None,
        tz: str | None = None,
    ) -> list[IMessageConversationSummary]:
        """List iMessage conversations with latest-message preview.

        Args:
            agent_identity_id: Narrow to one agent identity. Ignored
                for identity-scoped keys (always their own identity).
            limit: Max results to return (1–200).
            offset: Pagination offset.
            is_blocked: Tri-state filter applied to the underlying
                messages — ``True`` for only blocked, ``False`` for only
                non-blocked, ``None`` for all.
            include_groups: Include group conversations. Defaults to ``False``.
            start_datetime: Inclusive lower bound on ``created_at`` (str). Bare
                dates resolve to the start of that day; naive datetimes are
                interpreted in ``tz``; zoned datetimes are exact instants.
                ``None`` leaves the range open on this side.
            end_datetime: Upper bound on ``created_at`` (str). A bare date is
                whole-day inclusive. ``None`` leaves the range open.
            tz: IANA timezone name (str) governing zone-less values;
                ``None`` means UTC.
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if agent_identity_id is not None:
            params["agent_identity_id"] = str(agent_identity_id)
        if is_blocked is not None:
            params["is_blocked"] = is_blocked
        if include_groups:
            params["include_groups"] = True
        if start_datetime is not None:
            params["start_datetime"] = start_datetime
        if end_datetime is not None:
            params["end_datetime"] = end_datetime
        if tz is not None:
            params["tz"] = tz
        data = self._http.get("/conversations", params=params)
        return [IMessageConversationSummary._from_dict(c) for c in data]

    def get_conversation(
        self,
        conversation_id: UUID | str,
        *,
        agent_identity_id: UUID | str | None = None,
    ) -> IMessageConversation:
        """Get one iMessage conversation by ID.

        Args:
            conversation_id: UUID of the conversation.
            agent_identity_id: Optional identity assertion; 404s when
                the conversation belongs to a different identity.
        """
        params: dict[str, Any] = {}
        if agent_identity_id is not None:
            params["agent_identity_id"] = str(agent_identity_id)
        data = self._http.get(f"/conversations/{conversation_id}", params=params)
        return IMessageConversation._from_dict(data)

    def send_reaction(
        self,
        *,
        message_id: UUID | str,
        reaction: IMessageReactionType | str,
        part_index: int = 0,
    ) -> IMessageReaction:
        """Send a tapback reaction to a message.

        Args:
            message_id: UUID of the message being reacted to.
            reaction: Tapback kind. Sends accept the classic six;
                ``custom`` is inbound-only and rejected with 422.
            part_index: Part of a multi-part message to react to.
        """
        body: dict[str, Any] = {
            "message_id": str(message_id),
            "reaction": (
                reaction.value if isinstance(reaction, IMessageReactionType) else reaction
            ),
            "part_index": part_index,
        }
        data = self._http.post("/reactions", json=body)
        return IMessageReaction._from_dict(data)

    def mark_conversation_read(
        self,
        conversation_id: UUID | str,
    ) -> IMessageMarkReadResult:
        """Send a read receipt and mark inbound messages read locally.

        Args:
            conversation_id: UUID of the conversation.

        Returns:
            ``IMessageMarkReadResult`` with the count of rows updated.
        """
        data = self._http.post(
            "/mark-read",
            json={"conversation_id": str(conversation_id)},
        )
        return IMessageMarkReadResult._from_dict(data)

    def send_typing(self, conversation_id: UUID | str) -> None:
        """Show a typing indicator to the conversation's recipient.

        Args:
            conversation_id: UUID of the conversation.
        """
        self._http.post(
            "/typing",
            json={"conversation_id": str(conversation_id)},
        )

    def upload_media(
        self,
        *,
        content: bytes,
        filename: str,
        content_type: str | None = None,
    ) -> IMessageMediaUpload:
        """Upload media and get back a URL usable in ``media_urls``.

        Args:
            content: Raw file bytes (max 10 MiB).
            filename: Original filename, used for type inference.
            content_type: Optional MIME type; defaults server-side to
                ``application/octet-stream``.

        Returns:
            ``IMessageMediaUpload`` with the reusable ``media_url``.
        """
        data = self._http.post_multipart(
            "/media",
            field_name="file",
            filename=filename,
            content=content,
            content_type=content_type or "application/octet-stream",
        )
        return IMessageMediaUpload._from_dict(data)
